File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
import java.util.*;
class Solution {
    public int solution(String name) {
        int answer = 0;
        char[] array = name.toCharArray();
        // 상, 하 최적 이동 횟수
        for(int i = 0; i < array.length; i++) {
            if(array[i] == 'A') continue;
            answer += Math.min((array[i] - 'A'), ('Z' - array[i] + 1));
        }

        // 좌, 우 이동 최적값
        int len = array.length;
        int move = len - 1;
        for(int i = 0; i<len; i++){
            int next = i + 1;
            while(next < len && array[next] == 'A'){
                next++;
            }
            move = Math.min(move, i + len - next + Math.min(i, len-next));
        }
        answer += move;
        return answer;
    }
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <iostream>
#include <vector>

using namespace std;

int solution(string name) {
    int answer = 0;
    int name_len = name.length();
    // 상하 최적값
    for (const char& c : name)
    {
        if (c == 'A') continue;
        answer += min(c - 'A', 'Z' - c + 1);
    }
    // 좌우 최적값
    int move = name_len-1;
    for (int i = 0; i < name_len; i++)
    {
        int next = i+1;
        while (next < name_len && name[next] == 'A') next++;
        move = min(move, (i+name_len - next + min(i, name_len - next)));
    }
    answer += move;
    return answer;
}